src/modules/accounting/accounting.py

```diff
@@ -0,0 +1,222 @@
+from src.modules.accounting.bunker import BunkerService
+from src.typings import BlockStamp, Gwei
+from src.utils.abi import named_tuple_to_dataclass
+from src.web3py.typings import Web3
+
+
+logger = logging.getLogger(__name__)
+
+
+class Accounting(BaseModule, ConsensusModule):
+    CONSENSUS_VERSION = 1
+    CONTRACT_VERSION = 1
+
+    def __init__(self, w3: Web3):
+        self.report_contract = w3.lido_contracts.accounting_oracle
+        super().__init__(w3)
+        self.lido_validator_state_service = LidoValidatorStateService(self.w3)
+        self.bunker_service = BunkerService(self.w3)
+
+    # Oracle module: loop method
+    def execute_module(self, blockstamp: BlockStamp) -> None:
+        report_blockstamp = self.get_blockstamp_for_report(blockstamp)
+        if report_blockstamp:
+            self.process_report(report_blockstamp)
+
+            latest_blockstamp = self._get_latest_blockstamp()
+            if not self.is_extra_data_submitted(latest_blockstamp):
+
+                slots_to_sleep = self._get_slot_delay_before_data_submit(blockstamp)
+                chain_config = self._get_chain_config(blockstamp)
+                seconds_to_sleep = slots_to_sleep * chain_config.seconds_per_slot
+                logger.info({'msg': f'Sleep for {seconds_to_sleep} before sending extra data.'})
+                sleep(seconds_to_sleep)
+
+                self._submit_extra_data(report_blockstamp)
+
+    def _submit_extra_data(self, blockstamp: BlockStamp) -> None:
+        extra_data = self.lido_validator_state_service.get_extra_data(blockstamp, self._get_chain_config(blockstamp))
+
+        tx = self.report_contract.functions.submitReportExtraDataList(extra_data.extra_data)
+
+        if not variables.ACCOUNT:
+            logger.info({'msg': 'No account provided to submit extra data. Dry mode'})
+            return
+        if self.w3.transaction.check_transaction(tx, variables.ACCOUNT.address):
+            self.w3.transaction.sign_and_send_transaction(tx, variables.GAS_LIMIT, variables.ACCOUNT)
+
+    # Consensus module: main build report method
+    @lru_cache(maxsize=1)
+    def build_report(self, blockstamp: BlockStamp) -> tuple:
+        report_data = self._calculate_report(blockstamp)
+        logger.info({'msg': 'Calculate report for accounting module.', 'value': report_data})
+        return report_data.as_tuple()
+
+    # # Consensus module: if contract got report data
+    def is_main_data_submitted(self, blockstamp: BlockStamp) -> bool:
+        processing_state = self._get_processing_state(blockstamp)
+        logger.info({'msg': 'Check if main data was submitted.', 'value': processing_state.main_data_submitted})
+        return processing_state.main_data_submitted
+
+    # Consensus module: if contract could accept any sort of report
+    def is_contract_reportable(self, blockstamp: BlockStamp) -> bool:
+        is_reportable = not self.is_main_data_submitted(blockstamp) or not self.is_extra_data_submitted(blockstamp)
```

Comment:
**last_finalized_blockstamp** -> `execute_module` -> `get_blockstamp_for_report` -> `is_contract_reportable`

`is_contract_reportable` is called from `get_blockstamp_for_report` that is called with `last_finalized_blockstamp` in `execute_module`

in turn `is_contract_reportable` calls `is_main_data_submitted` and `is_extra_data_submitted` with `last_finalized_blockstamp` but i guess it should perform these checks on the latest block